### ISSUE TITLE
Provide an example of click for arg parsing

### ISSUE THREAD

**Issue by fjebaker** [edbennett/publishing-analysis#19]
[Click](https://click.palletsprojects.com/en/8.1.x/) is a small Python library for creating command line interfaces, which has numerous benefits over Python's argparse. 

To list a few:
- it's a modern tool,
- has modern and expressive syntax,
- more type support, including file paths and consistency checks,
- widely used,
- [improves upon argparse](https://click.palletsprojects.com/en/8.1.x/why/#why-not-argparse) in [multiple ways](https://click.palletsprojects.com/en/8.1.x/why/#why-no-auto-correction),
- creates a _consistent_ and familiar interface for end-users, and
- adds an additional dependency for later.

Adapting the example in [03 Documentation and automation](https://edbennett.github.io/publishing-analysis/03-documentation/index.html):
```py
import click

@click.command()
@click.option(
    "--infile", 
    type=click.Path(exists=True),
    help="input file name"
)
@click.option(
    "-n", "--num",
    type=int,
    default=None,
    help="Output only n most frequent words"
)
def main(infile, num):
    """Count the occurrences of all words in a text and write them to a CSV-file."""
    word_counts = count_words(infile)
    util.collection_to_csv(word_counts, num=num)
    # ...

if __name__ == "__main__":
    main()
```

**Comment by mmesiti** [comment]
I like click. Reasons not to use click:

A. Pedagogical:
  1. needs manual install: the audience of this lesson might be in the position to know/manage all of this (especially at the end of the reproducible environment section (https://edbennett.github.io/publishing-analysis/06-environments/index.html) but it adds another point of failure;
  2. one needs to use decorators, which can look a little mysterious at first.

B: Factual:
  1. it is controversial: http://xion.io/post/programming/python-dont-use-click.html and it might not be the best tool for the job (see 3);
  2. it is not as well known as argparse;
  3. it forces you into a particular structure of your script: with `argparse`, you just create a parser object that you can pass around or access its fields as you please. With `click`, you need to put all your functions in "main" function that should be called within the usual main block. This adds boilerplate code.

`click` is cool to create arbitrary trees of subcommands, which are now really fashionable. git, docker, poetry/pdm/pipenv/pytest do use subcommands nowadays, flask uses indeed click for that. I honestly find them easier to type than nested `--options`.

This is orthogonal to what we want to do in this section though: we want to create simple scripts that do one thing and can be controlled by a bash script.
Using click one can create nice interactive interfaces from a library of functions, but would still need the shell (or make) on top to automate it.

**Comment by fjebaker** [comment]
These are all fair points! 

To address a few things:

> needs manual install

Section [04 Jupyter notebooks (...)](https://edbennett.github.io/publishing-analysis/04-jupyter/index.html) already requires the audience to install `nbstripout` with pip prior to the environment lesson, but I accept this does add more complexity.

> one needs to use decorators, which can look a little mysterious at first

I don't think the use of decorators should be a reason to avoid using a library -- a user of click doesn't need to know the internals, and the API is written in a way where you really don't need to know what they do, as it effectively becomes "one function argument per `@click.option`". I'd even argue that providing some kind of introduction to decorators in this way would go some way to de-mystifying them, and even reduce the fear newcomers to Python may have when they see them. It's not like the audience is encountering them in the wild, but instead is being introduced to them in an environment where they serve a clear purpose, under the instruction of experts (i.e. the carpenters), and with a clear goal. In this context they can just be explained as fancy functions that add arguments (just like an argument parser).

> it is controversial

I read that blog post in full and I don't think the author's arguments are particularly valid (see the comment section of the blog post for some of the reasons).

> it is not as well known as argparse

This is absolutely true, and I can also understand that `argparse` is always available, whereas `click` is a dependency that requires (at the very least) a way of installing it.

> it forces you into a particular structure of your script

I'd argue `argparse`'s parser object ends up doing exactly the same thing, as you still need to create it and call `parser.parse_args()` somewhere, and there are more ways of doing that "wrong" or in a manner that makes extending the code unneccessarily complex (there are many SO questions like [this](https://stackoverflow.com/questions/27957373/python-import-and-initialize-argparse-after-if-name-main) with differing opinions of what is the "correct" practice). With `click`, there are at least clear entrypoints into the code which coherently follow style guides like [PEP8's conduct on imports at the top](https://peps.python.org/pep-0008/#imports) (I understand this point is a bit pedantic, but so are linters :p).

> This adds boilerplate code.

`argparse` has a lot of boilerplate code as well, which (in my experience, especially with newcomers) may be located in many different places. With `click`, the paradigm of keeping similar functions in the same place is enforced.

> Using click one can create nice interactive interfaces from a library of functions,

`click` can be used in many different ways, but I agree that it is principally designed for this. Again though, I don't think this should neccessarily be a reason to avoid the library. `click` could be seen as using a sledgehammer to crack a wallnut with all the things it supports and can do, however most of those things the user doesn't need to consider as the library has sensible defaults.

<hr>

Ultimately, I personally think using `click` invites a greater learning opportunity for the audience, and encourages "better" practices. I have enjoyably used `argparse` and `click` in a fair few projects before, but I'd recommend `click` over argparse because I believe it to be a more _modern feeling_ solution. I completely understand why you might avoid including it in this course, but I'd at least suggest having an aside or admonition with a `click` example, both to show that other solutions to argument parsing exist, but also for perhaps more experienced Python programmers in your audience to discover (in the same way [Make is mention in part 03](https://edbennett.github.io/publishing-analysis/03-documentation/index.html#make-it-better)).

**Comment by mmesiti** [comment]
Regarding

> it forces you into a particular structure of your script

Let me rephrase that - it forces you into a structure that would be quite different from alternatives. When using `argparse` you are not using a very different structure than when using `sys.argv`. It is really easy to transition from using `sys.argv` to `argparse`. You could even define two equivalent functions doing the same things transparently, one using `sys.argv` and another using `argparse`.

It is true that the transformation from a simple `argparse` example to a simple `click` example is trivial, but while argparse decouples the "getting the arguments" bit from the program logic, click forces you to define functions to work specifically with command line arguments.

Is less flexibility necessarily a bad thing? No, if we all agree on the "good way" to do things - that is, if we all have the same needs. See, e.g., PEP20 - line 13.

I do like click's *modern feeling*, and likely it's more PEP20-like than argparse. `click` is for sure worth a mention that might lead bored readers to interesting readings. One optional challenge might ask the reader to try and implement the same thing with `click`, pointing at the documentation.

Best practices evolve in time, and maybe in 5 years we will all find that `click` is much better for whatever reason. Likely, research software is not immune from FOTM phenomena, so what is taught as "best practices" now will likely change in the future (but hopefully not the underlying principles).
Yet, at the moment, https://merely-useful.tech/py-rse/index.html mentions `argparse` and not `click`, so just to stick to commonly widely accepted practices (since this is a course and not a blog post) I believe the main narration should stick with `argparse`, at the cost of potentially slowing the progress of the Human Race in mastering nature.

**Comment by fjebaker** [comment]
> One optional challenge might ask the reader to try and implement the same thing with `click`, pointing at the documentation.

I think this is a cool idea!

> Yet, at the moment, https://merely-useful.tech/py-rse/index.html mentions `argparse` and not `click`

And this I completely accept as to why it is neccessary to use `argparse`. Since the course introduction refers to this book, it makes absolute sense to follow the same tools, so that the audience members looking for more have familiarity. Maybe I'll hoist the issue to the book's repository instead ;)

 > I believe the main narration should stick with `argparse`, at the cost of potentially slowing the progress of the Human Race in mastering nature
 
![click_argparse](https://user-images.githubusercontent.com/11492844/161861407-fd920909-df9d-4d30-baf3-2880295bff42.png)

**Comment by edbennett** [comment]
Interesting discussion. I agree that it would be good to include as a callout or a challenge, but not to rewrite the main body to use `click`.

I don't use `click` myself (mostly by inertia rather than any value judgement on it), but I think `click` makes more sense if you're writing code for yourself, or for other people who care a lot about code. But the target audience for the code that might be released by those following this workshop will likely include mostly people either unfamiliar or relatively inexperienced with Python; in this case, seeing an explicit "this thing parses arguments" is I think more readable than a bunch of decorators with the relatively un-Googlable and non-self-describing "`click`". (I say this as the first time I encountered a codebase based on `click`, it took half an hour to work out where the entry point to the application was, as I didn't realise the `click` decorators everywhere were defining a CLI. Possibly this reflects more on me than on `click`, but it's an experience that stuck with me)

The other thing is that as you say, "using `click` invites a greater learning opportunity for the audience"—more to learn means more conceptual overhead. There's a finite number of new concepts and things one can wrap their brain around in a given timespan, and I suspect that this lesson is already at or over that limit.

**Comment by fjebaker** [comment]
> more to learn means more conceptual overhead. There's a finite number of new concepts and things one can wrap their brain around in a given timespan, and I suspect that this lesson is already at or over that limit

I agree with this too. You have managed to pack a great many very useful skills into this course, and our breakfast discussions of yesterday's material were extremely positive, with the course progressing at an approriate pace. 

>  say this as the first time I encountered a codebase based on click, it took half an hour to work out where the entry point to the application was, as I didn't realise the click decorators everywhere were defining a CLI.

I'd say this would be a good reason just to mention that `click` exists in this course, perhaps with a very small example, so that if the audience do encounter it in the wild they may remember seeing it before and have a point of reference for understanding the structure (again, in the same way Make is mentioned). Obviously there are a myriad of argument parsing libraries in Python, but I reckon `argparse` and `click` cover the extreme ends of the syntax spectrum, and are useful to _know_ about, if not directly learn at a later stage.

I'll amend the title of this issue to reflect the discussion :)
